src/account-db.js

```diff
@@ -40,50 +41,214 @@ function hashPassword(password) {
 
 export function needsBootstrap() {
   let accountDb = getAccountDb();
-  let rows = accountDb.all('SELECT * FROM auth');
-  return rows.length === 0;
+  let row = accountDb.first('SELECT count(*) FROM auth');
+  return row['count(*)'] === 0;
 }
 
-export function bootstrap(password) {
+// Supported login settings:
+// "password": "secret_password",
+// "openid": {
+//   "issuer": "https://example.org",
+//   "client_id": "your_client_id",
+//   "client_secret": "your_client_secret",
+//   "server_hostname": "https://actual.your_website.com"
+// }
+export function bootstrap(loginSettings) {
   let accountDb = getAccountDb();
 
-  let rows = accountDb.all('SELECT * FROM auth');
-  if (rows.length !== 0) {
+  if (!needsBootstrap()) {
     return { error: 'already-bootstrapped' };
   }
 
-  if (password == null || password === '') {
-    return { error: 'invalid-password' };
+  if (
+    !loginSettings.hasOwnProperty('password') &&
+    !loginSettings.hasOwnProperty('openid')
+  ) {
+    return { error: 'no-auth-method-selected' };
   }
 
-  // Hash the password. There's really not a strong need for this
-  // since this is a self-hosted instance owned by the user.
-  // However, just in case we do it.
-  let hashed = hashPassword(password);
-  accountDb.mutate('INSERT INTO auth (password) VALUES (?)', [hashed]);
+  if (loginSettings.hasOwnProperty('password')) {
+    const password = loginSettings.password;
+    if (password === null || password === '') {
+      return { error: 'invalid-password' };
+    }
+  }
 
-  let token = uuid.v4();
+  if (loginSettings.hasOwnProperty('openid')) {
+    const config = loginSettings.openid;
+    if (!config.hasOwnProperty('issuer')) {
+      return { error: 'missing-issuer' };
+    }
+    if (!config.hasOwnProperty('client_id')) {
+      return { error: 'missing-client-id' };
+    }
+    if (!config.hasOwnProperty('client_secret')) {
+      return { error: 'missing-client-secret' };
+    }
+    if (!config.hasOwnProperty('server_hostname')) {
+      return { error: 'missing-server-hostname' };
+    }
+    // Beyond verifying that the configuration exists, we do not attempt
+    // to check if the configuration is actually correct.
+    // If the user improperly configures this during bootstrap, there is
+    // no way to recover without manually editing the database. However,
+    // this might not be a real issue since an analogous situation happens
+    // if they forget their password.
```

Comment:
(This could also be written up as a feature request and handled in a future PR)

Reply:
This is a good idea although I'd say outside the scope of this PR. I'm wary of allowing a bypass like this since we support multiple authentication methods at once - if the user configures both openid and password, and only uses password to log in, this would allow an attacker to gain access by resetting the config.

Reply:
If users are now generally expected to self-host, I don't see a big need for this as the authorization config can simply be overwritten server-side.